File: Algorithms/src/medium/Way to Home.cpp
#include<bits/stdc++.h>
using namespace std;
#define mod 1000000007
const int N=1004;
long long int fact[N],invfact[N];
long long fastexp(int num,int p)
{
    if(p==0) return 1;
    long long res=fastexp(num,p/2);
    if(p%2==0)
        return res*res%mod;
    else
        return (((res*res)%mod)*num)%mod;
}
int main()
{
    //freopen("in10.txt","r",stdin);
    //freopen("out10.txt","w",stdout);
    int n,m;
    scanf("%d %d",&n,&m);
    int i;
    int h=n-1;
    int r=m-1;
    if(h+1<r)
        printf("0\n");
    else
    {
        fact[0]=1;
        invfact[0]=1;
        for(i=1;i<N;i++)
        {
            fact[i]=(fact[i-1]*i)%mod;
            invfact[i]=fastexp(fact[i],mod-2);
        }
        long long ans=(((fact[h+1]*invfact[r])%mod)*invfact[h+1-r])%mod;
        printf("%lld\n",ans);
    }
    return 0;
}

File: Algorithms/src/medium/way to home.java
import java.util.*;

public class Main {
    static final long MOD = 1000000007;
    static final int N = 1004;

    static long[] fact = new long[N];
    static long[] invfact = new long[N];

    static long fastExp(long num, long p) {
        if (p == 0) return 1;
        long res = fastExp(num, p / 2);
        if (p % 2 == 0)
            return (res * res) % MOD;
        else
            return (((res * res) % MOD) * num) % MOD;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int m = sc.nextInt();

        int h = n - 1;
        int r = m - 1;

        if (h + 1 < r) {
            System.out.println(0);
            return;
        }

        fact[0] = 1;
        invfact[0] = 1;

        for (int i = 1; i < N; i++) {
            fact[i] = (fact[i - 1] * i) % MOD;
            invfact[i] = fastExp(fact[i], MOD - 2);
        }

        long ans = (((fact[h + 1] * invfact[r]) % MOD) * invfact[h + 1 - r]) % MOD;

        System.out.println(ans);
    }
}
